flang/lib/Optimizer/Dialect/FIRType.cpp

```diff
@@ -330,6 +330,13 @@ bool isAllocatableOrPointerArray(mlir::Type ty) {
   return false;
 }
 
+bool isTypeWithDescriptor(mlir::Type ty) {
+  if (fir::isPointerType(ty) || fir::isAllocatableType(ty) ||
```

Comment:
It can probably be done with just `isa<BaseBoxType>(unwrapRefType(ty))`

Reply:
If you want to catch polymorphic entities you need to use Slava's suggestion.

Reply:
Thank you, I'll utilise the isa!